backend/aopp.go

```diff
@@ -0,0 +1,303 @@
+	aoppStateAwaitingKeystore aoppState = "awaiting-keystore"
+	// Keystore is registered - the user chooses an account from which to get a receive address
+	// from.
+	aoppStateChoosingAccount aoppState = "choosing-account"
+	// The account is still syncing - need to wait for that to finish before we can get a fresh
+	// address.
+	aoppStateSyncing aoppState = "syncing"
+	// The user is prompted to confirm and sign the address on the device.
+	aoppStateSigning aoppState = "signing"
+	// Everything went well, the address and signature was delievered to the AOPP callback.
+	aoppStateSuccess aoppState = "success"
+)
+
+// AOPP holds all the state needed to process an AOPP (Address Ownership Proof Protocol) request.
+type AOPP struct {
+	// State is the current state the request is in. See `aoppState*` for the possible values.
+	State aoppState `json:"state"`
+	// ErrorCode is an "aopp*" error code, see errors.go. Only applies if State == aoppStateError.
+	ErrorCode ErrorCode `json:"errorCode"`
+	// Accounts is the list of accounts the user can choose from. Only applies if State == aoppStateChoosingAccount.
+	Accounts []account `json:"accounts"`
+	// Address that will be delivered to the requesting party via the callback. Only applies if State == aoppStateSigning.
+	Address string `json:"address"`
+	// CallbackHost contains the host of the AOPP callback URL. Available for all states except
+	// aoppStateInactive.
+	CallbackHost string `json:"callbackHost"`
+	// coinCode is the requested asset. Available for all states except aoppStateInactive.
+	coinCode coinpkg.Code
+	// message is the requested message to be signed. Available for all states except
+	// aoppStateInactive.
+	message string
+	// callback is the AOPP callback param. Available for all states except aoppStateInactive.
+	callback string
+}
+
+// AOPP returns the current AOPP state.
+func (backend *Backend) AOPP() AOPP {
+	defer backend.accountsAndKeystoreLock.RLock()()
+	return backend.aopp
+}
+
+// notifyAOPP sends the aopp state to the frontend. `accountsAndKeystoreLock` must be held when
+// calling this function.
+func (backend *Backend) notifyAOPP() {
+	backend.Notify(observable.Event{
+		Subject: "aopp",
+		Action:  action.Replace,
+		Object:  backend.aopp,
+	})
+}
+
+// AOPPCancel resets the aopp state.
+func (backend *Backend) AOPPCancel() {
+	defer backend.accountsAndKeystoreLock.Lock()()
+	backend.aopp = AOPP{State: aoppStateInactive}
+	backend.notifyAOPP()
+}
+
+// aoppSetError pushes an error to the frontend to display. `accountsAndKeystoreLock` must be held
+// when calling this function.
+func (backend *Backend) aoppSetError(err ErrorCode) {
+	backend.aopp.State = aoppStateError
+	backend.aopp.ErrorCode = err
+	backend.notifyAOPP()
+}
+
+// aoppKeystoreRegistered must be called after a keystore is available, to display a list of
+// accounts to choose from. `accountsAndKeystoreLock` must be held when calling this function.
+func (backend *Backend) aoppKeystoreRegistered() {
```

Comment:
please add a comment that this is also called right away, "internally" by handleAOPP if a keystore is already available at the time AOPP is initiated.

Reply:
:heavy_check_mark: